profiler/src/main/java/com/splunk/opentelemetry/logs/BatchingLogsProcessor.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.logs;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * A log processor that batches logs until a max number are buffered or a time limit has passed,
+ * which ever comes first.
+ */
+public class BatchingLogsProcessor implements LogsProcessor {
+  private final int maxBatchSize;
+  private final Duration maxTimeBetweenBatches;
+  private final List<LogEntry> batch;
+  private final Consumer<List<LogEntry>> batchAction;
+  private final Object lock = new Object();
+  private WatchdogTimer watchdog;
+
+  public BatchingLogsProcessor(
+      Duration maxTimeBetweenBatches, int maxBatchSize, Consumer<List<LogEntry>> batchAction) {
+    this.maxTimeBetweenBatches = maxTimeBetweenBatches;
+    this.maxBatchSize = maxBatchSize;
+    this.batchAction = batchAction;
+    this.batch = new ArrayList<>(maxBatchSize);
+  }
+
+  public void start() {
+    synchronized (lock) {
+      if (watchdog != null) {
+        throw new IllegalStateException("Already running");
+      }
+      watchdog = new WatchdogTimer(maxTimeBetweenBatches, this::doAction);
+      watchdog.start();
+    }
+  }
+
+  void stop() {
+    synchronized (lock) {
+      if (watchdog == null) {
+        throw new IllegalStateException("Not running");
+      }
+      watchdog.stop();
+      watchdog = null;
+      doAction();
+    }
+  }
+
+  @Override
+  public void log(LogEntry log) {
+    synchronized (lock) {
+      batch.add(log);
+      if (batch.size() >= maxBatchSize) {
+        doAction();
```

Comment:
I don't think this is a good idea...to continue to block while the action runs in the same thread. At least, if this were in the SDK, I think we'd definitely need to make 100% sure that this didn't block the hot path.

Reply:
Oh yeah interesting. For this particular profiling use case, we are definitely not in a hot path and we can back up a little during export no problem. I can definitely see wanting to background the actual action/work.  I can think of a couple of options here:

* Always run the action in a background thread.  It's probably an extra/unnecessary/extra thread in the profiling use case, but should be fine.
* Keep the BatchingLogsProcessor like it is now, but clarify that action can be invoked from the calling thread if capacity is reached...and then build a backgrounding action to suppose the other use case.

I suppose I'm torn here.  The first option is probably more conventional, but I hate to fire up yet another thread when exporters themselves may be async/background. Maybe the second idea is just harder to understand as well.

Reply:
Ok, I took a stab at doing the first thing and just running the action on a background thread.  Lemme know what you think!  🤘🏻